### PR TITLE
fix: resolve issue when loading a world a second time

### PR DESCRIPTION
**Pull request by pollend**
looks like gestalt loads a whole new environment so any assets from one session are not compatible for another session. there are two main things I do, I wipe all the prefabs and let the engine re-resolve them when the world loads and unload any assets that are not loaded in by core.

### CODE REVIEW

File: engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java
Comment (pollend) [inline engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java:135]: umm must be a better way to clean this up :-1:

File: engine/src/main/java/org/terasology/engine/core/modes/StateLoading.java
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/modes/StateLoading.java:169]: Does this need to be in `initClient` as well?

File: engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java
Comment (keturn) [inline engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java:25]: What makes this better as a LoadProcess instead of a Subsystem?

more explicit ordering of load processes? or we need a different context here than was available for any of the subsystem lifecycle phases?

(and heck it is a lot of lines of code for these two expressions, either way)

**Review by keturn — CHANGES_REQUESTED** [review]
I think my main question here is: can we run the disposal code when we switch _out_ of the game environment? Wouldn't that be a better match to the expected lifecycle of these assets?

File: engine/src/main/java/org/terasology/engine/core/modes/StateLoading.java
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/core/modes/StateLoading.java:169]: not sure, I just managed to get this to work for single player still need to test things.

File: engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java:25]: we need a new rendering system when you restart a world.

Comment (jdrueckert) [reply engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java:25]: For my information: Why do we need a new rendering system when restarting a world?

File: engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java
Comment (jdrueckert) [inline engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java:133]: Why do we unregister prefab formats and dispose assets _at this point_?
Shouldn't this be done when exiting the previous world rather than when entering the next? :thinking:

Comment (skaldarnar) [reply engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java:133]: Same question as https://github.com/MovingBlocks/Terasology/pull/4829#pullrequestreview-714350912 :upside_down_face:

File: engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java:25]: its all module drive there really isn't a reason to keep it around as a singleton.

Comment (jdrueckert) [reply engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java:25]: Okay

**Comment by jdrueckert** [comment]
[Second World Start Bug](https://trello.com/c/iSiODJUC/145-second-world-start-bug)

**Review by keturn — APPROVED** [review]

File: engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java:25]: yea.. we do this for quite a few loading things.

File: engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java:133]: we end up loading those classes again on the classpath so the prefabs are not compatible between sessions.

**Comment by jdrueckert** [comment]
I tested this out and it definitely is not crashing anymore, neither when loading nor when starting a second or third world! :tada: 

_However_... I saw the following, so I assume we're still not cleaning up everything properly?
image

I was in my LaS workspace (because testing some other things before) and did the following:
1. Start a new LaS world
2. Load an older LaS world
3. Start a new Core Gameplay world

**Comment by pollend** [comment]
this only broke after the tinting was added I'm sure its a couple line fix but I thought it was out of the scope of this change. @jdrueckert I was planning on looking at this separately.

**Review by jdrueckert — COMMENTED** [review]
Tests out fine now and I don't see anything in the code that worries me. However, I'd like @keturn and / or @skaldarnar to have a final look, too, so I won't approve this yet.
@keturn, @skaldarnar feel free to approve and merge after you've had a look in case you're good with the changes.

File: engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/bootstrap/EnvironmentSwitchHandler.java:102]:
```suggestion
```

File: engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java:114]: For readability, let's do this in blocks, i.e., first we dispose `chunkProvider`, then `assetTypeManager`, ...

```suggestion
        chunkProvider.dispose();

        AssetTypeManager assetTypeManager = context.get(ModuleAwareAssetTypeManager.class);
```

Alternatively, separate the context retrieval from the dispose actions with a new line.

File: engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java:115]: What is `k` supposed to mean here? Can we please name it `assetType`?

File: engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java:123]: Please add a dosctring why we need to dispose these separately.

I guess this is because they are part of the engine module and are thus skipped by the statement above (https://github.com/MovingBlocks/Terasology/pull/4829/files#diff-b246f2c1e7c1a8f480631ab231b8a25b8b1b7bb5074d6b5b6ed3c6e01e44e0bcR117).

I like to have the info here about
- why do this in a separate statement
- why those two classes/how did we figure that out

File: engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java:125]: Please avoid more than one empty line
```suggestion

```

File: engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java:164]: Please avoid more than one empty line
```suggestion
```

File: engine/src/main/java/org/terasology/engine/world/block/DefaultColorSource.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/world/block/DefaultColorSource.java:20]: Is doing this lookup in `CoreRegistry` fast? I think @tolziplohu put this optimization on purpose to reduce the necessary lookups in the registry with each lookup of a color... :thinking:

File: engine/src/main/java/org/terasology/engine/world/block/loader/BlockFamilyDefinitionFormat.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/world/block/loader/BlockFamilyDefinitionFormat.java:321]: All these changes are unrelated to this PR. Please revert them.

File: engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java
Comment (skaldarnar) [reply engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java:25]: Can we please persist this information in a docstring. Probably the class doc here is a good place to capture this.

**Review by skaldarnar — CHANGES_REQUESTED** [review]

File: engine/src/main/java/org/terasology/engine/world/block/DefaultColorSource.java
Comment (pollend) [reply engine/src/main/java/org/terasology/engine/world/block/DefaultColorSource.java:20]: yea, but we can't save it as a static variable.

File: engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/modes/loadProcesses/InitialiseRendering.java:12]: @pollend does this description sound right to you?

```suggestion
/**
 * Add {@link RenderingModuleRegistry} to the game {@link Context}.
 * 
 * The rendering system is required whenever a client starts or joins a game. As rendering may fail to re-initialise
 * correctly when it has previously been constructed, this loading process will populate the {@link Context} with a
 * freshly created rendering system.
 * 
 * When switching the game state, the rendering system can just be disposed with the old state.
 */
```

File: engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java
Comment (skaldarnar) [inline engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java:121]:
```suggestion
        assetTypeManager.getAssetTypes().forEach(assetType -> {
            for (ResourceUrn urn : assetType.getLoadedAssetUrns()) {
                if (!urn.getModuleName().equals(TerasologyConstants.ENGINE_MODULE)) {
                    assetType.getAsset(urn).ifPresent(Asset::dispose);
                }
            }
        });
```

File: engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java
Comment (skaldarnar) [reply engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java:115]:
```suggestion
        // dispose all module assets
        assetTypeManager.getAssetTypes().forEach(k -> {
```

File: engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java
Comment (skaldarnar) [reply engine/src/main/java/org/terasology/engine/core/modes/StateIngame.java:123]:
```suggestion
        // dispose engine assets that should not be kept when switching game states
        assetTypeManager.getAssetType(BlockFamilyDefinition.class).ifPresent(AssetType::disposeAll);
        assetTypeManager.getAssetType(Prefab.class).ifPresent(AssetType::disposeAll);

```

**Review by skaldarnar — APPROVED** [review]

**Pull request merged**
